ros2topic/ros2topic/verb/delay.py

```diff
@@ -0,0 +1,236 @@
+
+    def __init__(self, node, window_size):
+        import threading
+        self.lock = threading.Lock()
+        self.last_msg_tn = 0
+        self.msg_t0 = -1.
+        self.msg_tn = 0
+        self.delays = []
+
+        # can't have infinite window size due to memory restrictions
+        if window_size < 0:
+            window_size = 50000
+        self.window_size = window_size
+
+        self._clock = node.get_clock()
+
+    def callback_delay(self, msg):
+        """
+        Calculate delay time.
+
+        :param msg: Message instance
+        """
+        if not hasattr(msg, 'header'):
+            raise RuntimeError('msg does not have header')
+        with self.lock:
+            curr_rostime = self._clock.now()
+
+            # time reset
+            if curr_rostime.nanoseconds == 0:
+                if len(self.delays) > 0:
+                    print('time has reset, resetting counters')
+                    self.delays = []
+                return
+
+            curr = curr_rostime.nanoseconds
+            if self.msg_t0 < 0 or self.msg_t0 > curr:
+                self.msg_t0 = curr
+                self.msg_tn = curr
+                self.delays = []
+            else:
+                curr_sec = curr_rostime.to_msg().sec
+                curr_nanosec = curr_rostime.to_msg().nanosec
+
+                delay = ((curr_sec + curr_nanosec * 1e-9) -
+                         (msg.header.stamp.sec + msg.header.stamp.nanosec * 1e-9))
+                self.delays.append(delay)
+                self.msg_tn = curr
+
+            if len(self.delays) > self.window_size:
+                self.delays.pop(0)
+
+    def get_delay(self):
+        """
+        Calculate the average publising delay.
+
+        :returns: tuple of stat results
+            (rate, min_delta, max_delta, standard deviation, window number)
+            None when waiting for the first message or there is no new one
+        """
+        if self.msg_tn == self.last_msg_tn:
+            return
+        with self.lock:
+            if not self.delays:
+                return
+            n = len(self.delays)
+
+            mean = sum(self.delays) / n
+            std_dev = math.sqrt(sum((x - mean)**2 for x in self.delays) / n)
+
+            max_delta = max(self.delays)
+            min_delta = min(self.delays)
+
+            self.last_msg_tn = self.msg_tn
+        return mean, min_delta, max_delta, std_dev, n
+
+    def print_delay(self):
+        """Print the average publishing delay to screen."""
+        if not self.delays:
+            return
+        ret = self.get_delay()
+        if ret is None:
+            print('no new messages')
+            return
+        delay, min_delta, max_delta, std_dev, window = ret
+        print('average delay: %.3f\n\tmin: %.3fs max: %.3fs std dev: %.5fs window: %s'
+              % (delay, min_delta, max_delta, std_dev, window))
+
+
+def _rostopic_delay(node, topic, window_size=-1, blocking=False):
```

Comment:
I would suggest making `window_size` a required argument - otherwise this function needs to pick an opaque default value.

How does `blocking=False` work if the `msg_class` can't be determined yet?

Reply:
This is original design. I think general user will not concern about window_size and just easy to use the tool. It will give user a little more trouble every time to type --window as required argument. It may also confuse user about the window meaning (first time I use rostopic, I really don't know what this window mean :))

Sorry but not considered when blocking=False, thanks for review, fixed by 0e79d06

Reply:
set blocking=True by default. commit again: c3e506f